### PR TITLE
#25522: Add fp32 support for tanhshrink

### LINKED ISSUE #25522: Add FP32 support for Tanhshrink
- [x] Add fp32 for tanhshrink
- [x] Make accurate version as default version for tanh, tanhshrink.

### PR DESCRIPTION
**Pull request by mouliraj-mcw**
### Ticket
Link to Github Issue #25522 

### What's changed
- Added fp32 support for tanhshrink.
- Modified ttnn.tanh flags: accurate version is now the default version, the faster LUT version is available by setting `approx=True`
- Removed accuracy flag for `ttnn.tanh` and added `approx` flag

### Perf results:

Accurate mode (Default version):

tanh -  5325 ns 
tanhshrink - 5454 ns 

Approx Mode:

tanh - 1789ns ( ~66% faster)
tanhshrink -  1874ns (~65% faster)

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16945288849)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16945299269)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 1]
```diff
@@ -295,9 +294,9 @@ Tensor Tanh::invoke(
     const Tensor& input_tensor,
     const std::optional<MemoryConfig>& memory_config,
     const std::optional<Tensor>& optional_output_tensor,
-    bool accuracy) {
+    bool approx) {
     UnaryOpType op_type = UnaryOpType::TANH;
-    if (!accuracy) {
+    if (approx) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp]: if (approx or dtype is block format )

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 2]
```diff
@@ -361,9 +360,15 @@ Tensor Tanhshrink::invoke(
     QueueId queue_id,
     const Tensor& input_tensor,
     const std::optional<MemoryConfig>& memory_config,
-    const std::optional<Tensor>& optional_output_tensor) {
+    const std::optional<Tensor>& optional_output_tensor,
+    bool approx) {
     UnaryOpType op_type = UnaryOpType::TANHSHRINK;
-    return detail::unary_impl(queue_id, input_tensor, {UnaryWithParam{op_type}}, memory_config, optional_output_tensor);
+    if (approx) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp]: if (approx or dtype is block format )

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp [hunk 3]
```diff
@@ -876,7 +876,7 @@ void bind_tanh(py::module& module, const unary_operation_t& operation) {
         Keyword Args:
             memory_config (ttnn.MemoryConfig, optional): Memory configuration for the operation. Defaults to `None`.
             output_tensor (ttnn.Tensor, optional): preallocated output tensor. Defaults to `None`.
-            accuracy (Boolean, optional): provides better accuracy for input range -3 to 3, for dtype BFLOAT16. Defaults to `False`.
+            approx (Boolean, optional): will be fast and approximate. Defaults to `False`.
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp]: approx (Boolean, optional): Enables a performance-optimized approximation method. When `True`, the operation runs faster but may produce results with minor precision differences. Defaults to `False`

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp [hunk 4]
```diff
@@ -895,29 +895,31 @@ void bind_tanh(py::module& module, const unary_operation_t& operation) {
                  - TILE
                  - 2, 3, 4
 
+            BFLOAT8_B is supported only if approx is True.
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp]: BFLOAT8_B/BFLOAT4_B is supported only for `approx=True` mode.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 5]
```diff
@@ -361,9 +360,15 @@ Tensor Tanhshrink::invoke(
     QueueId queue_id,
     const Tensor& input_tensor,
     const std::optional<MemoryConfig>& memory_config,
-    const std::optional<Tensor>& optional_output_tensor) {
+    const std::optional<Tensor>& optional_output_tensor,
+    bool approx) {
     UnaryOpType op_type = UnaryOpType::TANHSHRINK;
-    return detail::unary_impl(queue_id, input_tensor, {UnaryWithParam{op_type}}, memory_config, optional_output_tensor);
+    if (approx || input_tensor.dtype() == DataType::BFLOAT8_B) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp]: both block formats -  BFLOAT4_B

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 6]
```diff
@@ -295,9 +294,9 @@ Tensor Tanh::invoke(
     const Tensor& input_tensor,
     const std::optional<MemoryConfig>& memory_config,
     const std::optional<Tensor>& optional_output_tensor,
-    bool accuracy) {
+    bool approx) {
     UnaryOpType op_type = UnaryOpType::TANH;
-    if (!accuracy) {
+    if (approx || input_tensor.dtype() == DataType::BFLOAT8_B) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp]: both block formats -  BFLOAT4_B

File: ttnn/cpp/ttnn/operations/eltwise/unary/tanh_accurate/tanh_accurate.cpp [hunk 7]
```diff
@@ -33,9 +38,49 @@ Tensor Tanh_accurate::invoke(
...
+    bool fp32_dest_acc_en = preserve_fp32_precision or output_dtype == DataType::UINT32 or
+                            output_dtype == DataType::INT32 or output_dtype == DataType::FLOAT32 or
+                            input_dtype == DataType::UINT32 or input_dtype == DataType::INT32;
+
+    bool bfp8_pack_precise = input_dtype == DataType::BFLOAT8_B;
+
+    auto output_memory_config = optional_output_tensor.has_value()
+                                    ? optional_output_tensor.value().memory_config()
+                                    : memory_config.value_or(input_tensor.memory_config());
+
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/tanh_accurate/tanh_accurate.cpp]: instead of duplicating this code, have a common inline function that does this for you and calls prim::tanh_accurate with the parameters. both the ops can call this inline function

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp [hunk 8]
```diff
@@ -829,7 +829,7 @@ void bind_softplus(py::module& module, const unary_operation_t& operation) {
                * - Dtypes
                  - Layouts
                  - Ranks
-               * - BFLOAT16
+               * - BFLOAT16, BFLOAT8_B
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp:832]: FLOAT32 ?

**Review by KalaivaniMCW — APPROVED** [review]
Looks Good.
Please check pipelines carefully since we're making the accurate version to be the default. 
Check Single card, TG and T3K pipelines.

**Review by mcw-vishal — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    init_sfpu(cb_input, cb_output);
+
+    for (uint32_t block_index = 0; block_index < per_core_block_cnt; block_index++) {
+        cb_reserve_back(cb_output, per_core_block_dim);
+        for (uint32_t tile_index = 0; tile_index < per_core_block_dim; ++tile_index) {
+            cb_wait_front(cb_input, 1);
+            tile_regs_acquire();
+
+            // Pop tile after tile, copy to DST and pack
+            copy_tile_to_dst_init_short(cb_input);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: One minor thing I would suggest is changing this to `copy_tile_init(cp_input);`. It's defined as
```cpp
ALWI void copy_tile_init(uint32_t cbid) { copy_tile_to_dst_init_short(cbid); }
```
For simplicity I think it's better to avoid using `copy_tile_to_dst_init_short` directly unless you need to override the default argument `transpose = 0` (which you're not here).

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp [hunk 10]
```diff
@@ -876,7 +876,7 @@ void bind_tanh(py::module& module, const unary_operation_t& operation) {
         Keyword Args:
             memory_config (ttnn.MemoryConfig, optional): Memory configuration for the operation. Defaults to `None`.
             output_tensor (ttnn.Tensor, optional): preallocated output tensor. Defaults to `None`.
-            accuracy (Boolean, optional): provides better accuracy for input range -3 to 3, for dtype BFLOAT16. Defaults to `False`.
+            approx (Boolean, optional): Enables a performance-optimized approximation method. When True, the operation runs faster but may produce results with minor precision differences. Defaults to `False`.
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp]:
```suggestion
            fast_and_approximate_mode (Boolean, optional): Enables a performance-optimized approximation method. When True, the operation runs faster but may produce results with minor precision differences. Defaults to `False`.
```

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp [hunk 11]
```diff
@@ -895,29 +895,31 @@ void bind_tanh(py::module& module, const unary_operation_t& operation) {
                  - TILE
                  - 2, 3, 4
 
+            BFLOAT8_B/BFLOAT4_B is supported only for approx=True mode.
+
         Example:
             >>> tensor = ttnn.from_torch(torch.tensor([[1, 2], [3, 4]], dtype=torch.bfloat16), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
-            >>> output = {1}(tensor, accuracy=False)
+            >>> output = {1}(tensor, approx=False)
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.cpp]:
```suggestion
            >>> output = {1}(tensor, fast_and_approximate_mode=False)
```

**Review by patrickroberts — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            copy_tile(cb_input, 0, 0);
+            sub_binary_tile_init();
+            sub_binary_tile(0, 1);
+
+            tile_regs_commit();
+
+            tile_regs_wait();
+
+            pack_tile(0, cb_output);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: I think you could avoid the `copy_dest_values` if you did this:
```suggestion
            copy_tile_init(cb_input);
            copy_tile(cb_input, 0, 1);
            sub_binary_tile_init();
            sub_binary_tile(1, 0);

            tile_regs_commit();

            tile_regs_wait();

            pack_tile(1, cb_output);
```

File: tests/ttnn/unit_tests/operations/eltwise/test_tanh_accuracy.py [hunk 13]
```diff
@@ -61,15 +61,15 @@ def test_tanh_range(device, torch_dtype, ttnn_dtype):
...
 
     output_tensor = ttnn.to_torch(output_tensor)
 
     assert_allclose(output_tensor, torch_output_tensor, rtol=1e-05, atol=0.012)
     pcc, pcc_msg = assert_with_pcc(torch_output_tensor, output_tensor, 0.9999)
-    # pcc_msg 0.9999663646890817, accuracy=False pcc 0.9978378297942829
+    # pcc_msg 0.9999663646890817, fast_and_approximate_mode=True pcc 0.9978378297942829
     # pcc_msg 0.9999583453515977 - fpu arithmetic, pcc_msg 0.9999669593009368 sfpu arithmetic
-    # fp32 pcc_msg 0.9999829606828651 (accuracy=True) , 0.9977552960423647 (accuracy=False)
+    # fp32 pcc_msg 0.9999829606828651 (fast_and_approximate_mode=False) , 0.9977552960423647 (fast_and_approximate_mode=True)
```
Comment (dchenTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_tanh_accuracy.py:72]: nit: can you add a comment about roughly the performance tradeoff? so that information is complete. I guess this is the place people can get the idea of the two modes.

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp (hunk 9 above)
Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: Updated.

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp (hunk 12 above)
Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: We can’t implement this approach in SFPU because the destination index acts as the base, and it uses the destination offset to calculate the next tile starting from dst[0]. However, for FPU, you can simply swap the CB order.

Comment (patrickroberts) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: @mouliraj-mcw can you elaborate? Here's my understanding:
```cpp
copy_tile(cb_input, 0, 1); // cb_input[0] -> dst[1]
sub_binary_tile(1, 0); // dst[1] - dst[0] -> dst[1]
pack_tile(1, cp_output/*, output_tile_index = 0 */); // dst[1] -> cb_output[0]
```
I'm not sure why you say you can't implement this approach in SFPU. Have you actually tried the suggestion?

Comment (patrickroberts) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: Alternatively, you could perform `tanh_tile` on `dst[1]` and remove the `copy_dest_values` as well

Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: > @mouliraj-mcw can you elaborate? Here's my understanding:
> 
> ```c++
> copy_tile(cb_input, 0, 1); // cb_input[0] -> dst[1]
> sub_binary_tile(1, 0); // dst[1] - dst[0] -> dst[1]
> pack_tile(1, cp_output/*, output_tile_index = 0 */); // dst[1] -> cb_output[0]
> ```
> 
> I'm not sure why you say you can't implement this approach in SFPU. Have you actually tried the suggestion?

The feedback was to use sub_tile(1,0) and pack_tile(1) instead of sub_tile(0,1) and pack_tile(0).
However, this didn’t work as expected it gives input as output. I tried keeping sub_tile(1,0) while still using pack_tile(0), but the output I got had inverted signs.

Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: > Alternatively, you could perform `tanh_tile` on `dst[1]` and remove the `copy_dest_values` as well

Updated . This approach works.

Comment (patrickroberts) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/compute/tanhshrink_sfpu_kernel.cpp]: Interesting, thanks for that elaboration, I apologize for the confusion and I appreciate your patience with me. Glad the other suggestion worked.

**Review by patrickroberts — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

**Review by mtairum — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]
